src/pageObject.ts

```diff
@@ -0,0 +1,151 @@
+     */
+    as(component: new () => void) {
+        this.component = component;
+        return this;
+    }
+}
+
+export type NativeSelectorParams = {
+    driver: WebdriverIO.Browser;
+    parent: ChainablePromiseElement;
+    argument: string;
+};
+
+/**
+ * Define selector
+ */
+export interface LocatorDefinition {
+    (selector: any): Selector;
+
+    /**
+     * Define selector as a template
+     * @param {(argument: string) => string} selector - selector template
+     */
+    template: (selector: (argument: string) => string) => Selector;
+
+    /**
+     * Define selector using native wdio API
+     * @param {(argument: string) => string} selector - selector function
+     */
+    native: (selector: (params: NativeSelectorParams) => ChainablePromiseElement) => Selector;
+}
+
+export const locator: LocatorDefinition = function locator(selector: any): Selector {
+    return new Selector(selector);
+}
+
+locator.template = function(selector: (argument: string) => string) {
+    return new Selector(selector, 'template');
+}
+
+locator.native = function(selector: (params: NativeSelectorParams) => ChainablePromiseElement) {
+    return new Selector(selector, 'native');
+}
+
+export class ChainItem {
+    alias: string;
+    argument?: string;
+    selector: any;
+    type: string;
+
+    constructor({ alias, argument, selector, type }: { alias: string, argument?: string, selector?: string, type: string }) {
+        this.alias = alias;
+        this.argument = argument;
+        this.selector = selector;
+        this.type = type;
+    }
+}
+
+export function query(root: any, path: string) {
+    const elements = path.split(/\s*>\s*/);
+    const tokens = [];
+    let currentComponent = new root();
+    let currentAlias = 'root';
+    for (const element of elements) {
+        const groups = element.match(/^(?<alias>.+?)(?:\((?<argument>.+)\))?$/)?.groups as { alias: string, argument: string };
+        const alias = groups.alias.replace(/\s/g, '');
+        const currentElement = currentComponent[alias];
+        if (!currentElement) throw new Error(`Alias '${alias}' has not been found in '${currentAlias}'`);
+        currentAlias = groups.alias;
+        currentComponent = currentElement.component ? new currentElement.component() : null;
+
+        tokens.push(new ChainItem({
+            alias,
+            argument: groups.argument,
+            selector: currentElement.selector,
+            type: currentElement.type,
+        }));
+    }
+
+    return tokens;
+}
+
+export interface Locator {
+    (): ChainablePromiseElement;
+    collection: () => ChainablePromiseArray;
+};
+
+export function element(this: any, path: string): Locator {
+    const chain = query(this.config.pageObject, path);
+    const driver = this.wdio.driver as WebdriverIO.Browser;
+    const getter: Locator = function () {
+        let current = driver as unknown as ChainablePromiseElement;
+        for (const item of chain) {
+            switch (item.type) {
```

Comment:
should we add a `default` case for that switch case? Or is it possible that `item.type` will return some unexpected value?